### PR TITLE
NIFI-12078: Add JDBC Catalog Controller Service for Iceberg processors

### PR DESCRIPTION
**Pull request by mark-bathori**
# Summary

[NIFI-12078](https://issues.apache.org/jira/browse/NIFI-12078)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-common/pom.xml [hunk 1]
```diff
@@ -86,6 +89,21 @@
...
+        <dependency>
+            <groupId>org.apache.iceberg</groupId>
+            <artifactId>iceberg-azure</artifactId>
+            <version>${iceberg.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.iceberg</groupId>
+            <artifactId>iceberg-gcp</artifactId>
+            <version>${iceberg.version}</version>
+        </dependency>
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-common/pom.xml:106]: @mark-bathori Are these additional dependencies required as part of this change? They appear to be unrelated, and introduce a potentially large set of additional transitive dependencies.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.List;
+
+import static org.apache.nifi.services.iceberg.IcebergCatalogProperty.FILE_IO_IMPLEMENTATION;
+import static org.apache.nifi.services.iceberg.IcebergCatalogProperty.CLIENT_POOL_SERVICE;
+import static org.apache.nifi.services.iceberg.IcebergCatalogProperty.CLIENT_POOL_SIZE;
+import static org.apache.nifi.services.iceberg.IcebergCatalogProperty.WAREHOUSE_LOCATION;
+
+@Tags({"iceberg", "catalog", "service", "jdbc"})
+@CapabilityDescription("Catalog service using relational database to manage Iceberg tables through JDBC.")
+public class JdbcCatalogService extends AbstractCatalogService {
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java:38]: For naming purposes, I recommend `DatabaseCatalogService` instead of `Jdbc`, as JDBC identifies the Java API, versus Database, which identifies the type of implementation.
```suggestion
public class DatabaseCatalogService extends AbstractCatalogService {
```

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            .defaultValue("2")
+            .required(true)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor FILE_IO_IMPL = new PropertyDescriptor.Builder()
+            .name("File IO Implementation")
+            .description("Specifies the implementation of FileIO interface to be used. " +
+                    "The provided implementation have to include the class and full package name.")
+            .defaultValue("org.apache.iceberg.hadoop.HadoopFileIO")
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java:68]: Instead of a hard link to the class name, I recommend an `enum` of supported values with more generic names, such as `Hadoop File IO`. That allows the property value to change class names, if needed, without impacting the configuration. How are other implementation classes supported?

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            .required(true)
+            .build();
+
+    public static final PropertyDescriptor CONNECTION_POOL_SIZE = new PropertyDescriptor.Builder()
+            .name("Connection Pool Size")
+            .description("Specifies the size of the JDBC Connection Pool.")
+            .defaultValue("2")
+            .required(true)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .build();
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java:62]: Is this property necessary? It should be a function of the configuration for the Connection Pool itself.

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-common/pom.xml (hunk 1 above)
Comment (mark-bathori) [reply nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-common/pom.xml:106]: These dependencies contains different cloud provider IO implementations: S3FileIO, GCSFileIO and ADLSFileIO

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java (hunk 2 above)
Comment (mark-bathori) [reply nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java:38]: I wanted to stay consistent with the iceberg catalog type name which is JDBC Catalog (https://iceberg.apache.org/docs/1.5.0/jdbc/#jdbc-catalog). I'm not really sure if I rename it to Database catalog that it will be clear for the users that it means JDBC Catalog.

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java (hunk 3 above)
Comment (mark-bathori) [reply nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java:68]: Currently we have the HadoopFileIO and 3 cloud specific IO implementation (S3, ADLS, GCS) so I think we can make a fix list as you suggested. Custom FileIO implementation can be made but I think it's enough to support these.

File: nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java (hunk 4 above)
Comment (mark-bathori) [reply nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-services/src/main/java/org/apache/nifi/services/iceberg/JdbcCatalogService.java:62]: This property can be removed.

**Comment by mattyb149** [comment]
+1 LGTM, verified the expected behavior with PutIceberg. Thanks for the new feature! Merging to main
